File: tokenizer.cpp
#include <string> 
#include <vector> 
#include <sentencepiece_processor.h> 
#include <stdexcept>

class Tokenizer {
private: 
    sentencepiece::SentencePieceProcessor sp; 
    int bos_id, eos_id, pad_id, unk_id; 
    int max_len = 8192; 
    std::string model_path; 

public: 
    Tokenizer(const std::string& model_path) : model_path(model_path) { 
        auto status = sp.Load(model_path); 
        if (!status.ok()) { 
            throw std::runtime_error("Failed to load tokenizer model"); 
        }
        bos_id = sp.PieceToId("<bos>"); 
        eos_id = sp.PieceToId("<eos>"); 
        pad_id = sp.PieceToId("<pad>"); 
        unk_id = sp.unk_id(); 
    }

    std::vector<std::vector<int>> encode(
        const std::vector<std::string>& id, 
        bool special_tokens = true, 
        bool padding = false, 
        int max_len_override = -1
    )  {
        const int effective_max_len = (max_len_override == -1) ? max_len : max_len_override;
        
        std::vector<std::vector<int>> encoded_input; 

        for (const auto& t : id) { 
            std::vector<int> ids; 
            sp.Encode(t, &ids); 
            
            if (special_tokens) { 
                ids.insert(ids.begin(), bos_id); 
                ids.push_back(eos_id); 
            }
            
            if (padding && static_cast<int>(ids.size()) < effective_max_len) { 
                int padding_needed = effective_max_len - static_cast<int>(ids.size()); 
                ids.insert(ids.end(), padding_needed, pad_id);
            }
            
            encoded_input.push_back(ids); 
        }
        return encoded_input; 
    }
    
};

File: tokenizer.py
from typing import List, Union, Sequence, Optional
import sentencepiece as spm
import os

#Constants 
max_len = 8192
vocab_size = 262144
n_layers = 35
model_dim = 2048
ple_dim = 256

class Tokenizer: 
    
    def __init__(self, model_path: str): 
        if not os.path.exists(model_path):
            raise FileNotFoundError(f"Model file not found: {model_path}")
        self.model_path = model_path
        self.sp = spm.SentencePieceProcessor(model_file=self.model_path)
        self.bos_id = self.sp.piece_to_id("<bos>")
        self.eos_id = self.sp.piece_to_id("<eos>")
        self.pad_id = self.sp.piece_to_id("<pad>")
        self.unk_id = self.sp.unk_id()

    def encode(
        self, 
        text: Union[str, Sequence[str]], 
        special_tokens: bool = True, 
        max_len: Optional[int] = None, 
        padding: bool = False
    ) -> Union[List[int], List[List[int]]]:
        """
        Returns the python list of token ids
        """
        if max_len is None:
            max_len = self.max_len if hasattr(self, 'max_len') else 8192
            
        if isinstance(text, str): 
            text = [text]
            single_input = True 
        else:
            text = list(text)
            single_input = False 
        
        encoded_input: List[List[int]] = []

        for t in text: 
            ids: List[int] = list(self.sp.encode(t, out_type=int))

            if special_tokens: 
                ids = [self.bos_id] + ids + [self.eos_id]
            
            if padding and len(id) < max_len: 
                id.extend([self.pad_id] * (max_len) - len(id))

            encoded_input.append(ids)

        return encoded_input[0] if single_input else encoded_input